From: Stefan Becker <chemobejk@gmail.com>
Date: Sat, 09 May 2015 14:17:25 +0000
Subject: Add session management for KMainWindow
X-Git-Url: http://quickgit.kde.org/?p=kxmlgui.git&a=commitdiff&h=fa86f6e4afd4a4f32d297f271d3daececc6c0ba4
---
Add session management for KMainWindow

- replace the KConfig object in KConfigGui with a new one using the
session id and key from the QSessionManager object
- make sure that the data is stored after the application has update it
- if the KConfig object is represented by a local file then add a
discard command to the session manager object

BUG: 346768
REVIEW: 123706
---


--- a/src/kmainwindow.cpp
+++ b/src/kmainwindow.cpp
@@ -127,8 +127,10 @@
 {
 }
 
-bool KMWSessionManager::saveState(QSessionManager &)
-{
+bool KMWSessionManager::saveState(QSessionManager &sm)
+{
+    KConfigGui::setSessionConfig(sm.sessionId(), sm.sessionKey());
+
     KConfig *config = KConfigGui::sessionConfig();
     if (KMainWindow::memberList().count()) {
         // According to Jochen Wilhelmy <digisnap@cs.tu-berlin.de>, this
@@ -144,6 +146,19 @@
 
     KConfigGroup group(config, "Number");
     group.writeEntry("NumberOfWindows", n);
+
+    // store new status to disk
+    config->sync();
+
+    // generate discard command for new file
+    QString localFilePath =  QStandardPaths::writableLocation(QStandardPaths::GenericConfigLocation) + QLatin1Char('/') + config->name();
+    if (QFile::exists(localFilePath)) {
+        QStringList discard;
+        discard << QLatin1String("rm");
+        discard << localFilePath;
+        sm.setDiscardCommand(discard);
+    }
+
     return true;
 }
 

